### ISSUE TITLE
Advanced Font rendering for ocropus-linegen

### ISSUE THREAD

**Issue by wrznr** [tmbdev/ocropy#198]
## Expected Behavior
The text rendering by ocropus-linegen should be sensitive to RTL languages (scripts) and complex diacritics (i.e. sequences of diacritic marks).

## Current Behavior
Using the text rendering mechanics provided by PIL results in reversed images for RTL scripts and sequential placement of diacritic sequences (This also depends on the corresponding font. For a small number of fonts, e.g. DejaVu Sans, the placement is correct).

## Possible Solution
Using Cairo + Pango offers much better font rendering. A python wrapper is available (https://cairographics.org/pycairo/). The downside are additional dependencies for a feature which may not be interesting for many users. I file this issue to discuss the pros and cons of a more advanced approach to font rendering.

## Steps to Reproduce (for bugs)
sftp://lal.dwds.de/home/kmw/projects/dta/ocropus/iliad/devel/usr_share_fonts_truetype_culmus_FrankRuehlCLM-Medium/010000.gt.txt

1. Download gt file.
2. Install a hebrew-capable font (e.g. via the culmus package)
3. ocropus-linegen -e med -t 010000.gt.txt -o . -f /usr/share/fonts/truetype/culmus/FrankRuehlCLM-Medium.ttf

Result:

![](sftp://lal.dwds.de/home/kmw/projects/dta/ocropus/iliad/010000.bin.png)

4. Test the modified ocropus-linegen (for convenience, I kept the current version as comments and placed the modified code directly below each comment.

sftp://lal.dwds.de/home/kmw/built/ocropy/ocropus-linegen

Result:

![](sftp://lal.dwds.de/home/kmw/projects/dta/ocropus/iliad/010000.new.bin.png)

## Your Environment
* Python version: 2.7.3
* Git revision of ocropy:  commit fc525cb2fc07d7d8e29bbcf9fa3b7989d59df8cc, Merge: 4efbddc ae763b0
* Operating System and version: Debian 7.9

**Comment by amitdo** [comment]
See also:
https://github.com/mittagessen/kraken/blob/master/kraken/linegen.py

**Comment by wrznr** [comment]
Cool. I missed that one. Thanks.

**Comment by zuphilip** [comment]
Is this more or less a complete rewrite of the `linegen`-script here? Or can we easily copy something to improve the version here?

**Comment by amitdo** [comment]
>Is this more or less a complete rewrite of the linegen-script here?

It look like a complete rewrite.

**Comment by wrznr** [comment]
I would not say that it is a complete rewrite. To get Cairo + Pango-based font rendering running you would have to add a few imports and the modified getline function posted above. @mittagessen did a lot more in kraken's overhauled linegen script but the actual font rendering boils down to the steps sketched above.

**Comment by mittagessen** [comment]
The font rendering is a complete rewrite and doesn't use pycairo but accesses the libraries directly using ctypes. That's somewhat important because the official bindings are not installable through pip (requiring virtualenvs with global access) and are/were missing in Debian stable repositories. When I tried them they also were kind of buggy. Still, pulling _draw_on_surface and the appropriate ctypes boiler plate would be possible. An important part missing in the above code is setting the language on the layout which will affect rendering.

The noise model is also different, although in my experience neither work too well except possibly for boosting training data.

On a final note the typesetting produced by pango is still sub-par. The output will look different depending on the Unicode normalization of the input and the font used, e.g. NFD polytonic Greek text will have incorrectly placed diacritics with most fonts. My hackish solution is adding a switch to renormalize ground truth to another NF before rendering to `ketos linegen` but a proper workaround would entail looking at the actual font and how code points are mapped to graphemes. I had a discussion on the pango/harfbuzz lists about this and they agree it's undesired behavior but not something they are going to fix anytime soon.

**Comment by zuphilip** [comment]
Okay, thank for all the information. AFAIK kraken's version of linegen solves this issue (and maybe more) and I am okay with some limitations of the ocropy's version.

However, if someone wants to improve the ocropy version then let me know beforehand especially if any new dependencies are required.

Otherwise I will close the issue here.

**Comment by amitdo** [comment]
>The noise model is also different, although in my experience neither work too well except possibly for boosting training data.

l don't understand what you mean here. Can you further explain?

**Comment by mittagessen** [comment]
Sorry, the term boosting is unclear, augmentation is more appropriate. From my experience training on purely artificial data will result in unusably bad accuracy on real world data (<80%) even if typefaces are superficially similar. I am unsure about the source of this mallearning: could be either typesetting differences or issues with the noise model. On the other hand I've seen at least one publication using artificially created data (almost certainly from ocropy's linegen) in conjunction with natural data improving accuracy.

In fact this reminds me that I wanted to print, bind, and redigitize some texts in an attempt to isolate the error source.

**Comment by amitdo** [comment]
How many fonts you used? Ray Smith said he used >4000 digital fonts in Tesseract 4.00 for Latin script based languages. Huge number of fonts! Maybe he count any form of degradation as separate font.

**Comment by mittagessen** [comment]
Those were single-font tests with polytonic Greek fonts from the Greek Font Society (fairly similar to 18-19th century printing I had plenty of data for) and the Amiri font for classical Arabic. I would've expected some accuracy penalty as typically seen with slightly mismatching fonts (2-5%) but not a drop of more than 20% when evaluating on real world data.

Tesseract before 4.00 is less font-specific than ocropy although baseline accuracy is a lot lower especially without the language model. I believe Google tried to train generalized models for 4.00 that perform equally well on a large number of typographically diverse inputs so it makes sense to get as much training data as possible. It might have worked (tesseract bugged out on me the last time I tried it) although in my experience training for more than one typeface is quite finicky and creating a completely new model or "retargeting" an existing one with a smaller training set will reduce error reliably.

**Comment by wrznr** [comment]
I am okay with closing this issue since `ketos linegen` is available.

**Comment by ChillarAnand** [comment]
@zuphilip  Couple of days back Pillow added support for CTL(Complex Text Layout). I would recommend fixing linegen as it makes it easier to generate training data for many languages without any external dependencies.

**Comment by ChillarAnand** [comment]
Looks like there is nothing to do. With latest pillow ocropy works with CTL scripts.

**Comment by amitdo** [comment]
@ChillarAnand,

Thanks for the info about Pillow.

**Comment by amitdo** [comment]
https://github.com/python-pillow/Pillow/pull/2284

This feature is supported since release 4.2.0
https://github.com/python-pillow/Pillow/blob/master/CHANGES.rst

**Comment by Shreeshrii** [comment]
I tried out linegen yesterday. 

* Ocropy

ocropus-linegen is not rendering Devanagari script (Complex Indic script used for Hindi, Sanskrit, Marathi, Nepali etc) incorrectly. I have not tested other languages. 

I reinstalled Pillow and now have version 7.0.0 installed.

```
 ocropus-linegen  -f /home/ubuntu/.fonts/myfonts/Siddhanta.ttf -t test.txt -m 6
fonts ['/home/ubuntu/.fonts/myfonts/Siddhanta.ttf']
test.txt
# reading test.txt
got 6 lines
got 6 unique lines
base linegen
=== linegen/home_ubuntu_.fonts_myfonts_Siddhanta /home/ubuntu/.fonts/myfonts/Siddhanta.ttf
 0.50  0.50  58 भी इस समय दोनों की दोनों जिस रास्ते पर जा रही हैं, उसे देखते हुए इन दोनों में मेल और
 0.50  0.50  68 उन्नति और विकास के मार्ग में बाधक नहीं होना चाहते । उन्होंने यह्‌ मान लिया है कि
 0.50  0.50  68 भी इस समय दोनों की दोनों जिस रास्ते पर जा रही हैं, उसे देखते हुए इन दोनों में मेल और
 0.50  0.50  52 देश में ऐसे आदमियों की संख्या कम नहीं है जो उर्द्‌ और हिंदी की अलग अलग और स्वतंत्र
 0.50  0.50  46 फ़ारसी और अरबी के साथ स्वाभाविक सम्बंध है । और हिंदी का संस्कृत तथा प्राकृत के साथ
 0.50  0.50  62 देश में ऐसे आदमियों की संख्या कम नहीं है जो उर्द्‌ और हिंदी की अलग अलग और स्वतंत्र
```

![010001 bin](https://user-images.githubusercontent.com/5095331/73904672-beca3680-48c2-11ea-939c-c015a230b952.png)

उन्नति और विकास के मार्ग में बाधक नहीं होना चाहते । उन्होंने यह्‌ मान लिया है कि

* Kraken

ketos linegen generated the image correctly.

```
ketos linegen -ur NFC -u NFD -n 1500  --disable-degradation   -f Siddhanta test.txt
Read 6 unique lines
Σ (len: 48)
Symbols: ,अआइउएऐऔकखगचछजतथदधनपफबभमयरलवशषसह।५
Combining Characters: SPACE, DEVANAGARI SIGN ANUSVARA, DEVANAGARI SIGN NUKTA, DEVANAGARI VOWEL SIGN AA, DEVANAGARI VOWEL SIGN I, DEVANAGARI VOWEL SIGN II, DEVANAGARI VOWEL SIGN U, DEVANAGARI VOWEL SIGN UU, DEVANAGARI VOWEL SIGN VOCALIC R, DEVANAGARI VOWEL SIGN E, DEVANAGARI VOWEL SIGN AI, DEVANAGARI VOWEL SIGN O, DEVANAGARI SIGN VIRAMA, ZERO WIDTH NON-JOINER
```

![000001](https://user-images.githubusercontent.com/5095331/73904768-1c5e8300-48c3-11ea-8dda-f18890cfbae0.png)

उन्नति और विकास के मार्ग में बाधक नहीं होना चाहते । उन्होंने यह्‌ मान लिया है कि

**Comment by wrznr** [comment]
@Shreeshrii I also used `ketos linegen` successfully for Hebrew with Nikkud. It is the best choice for generating synthetic training materials.

**Comment by Shreeshrii** [comment]
@wrznr Thanks! That's good to know. I haven't tested RTL languages withe linegen yet.
